VideoModule to VideoBlock [SE-602]

symbolist opened the pull request:
Convert Video XModule to an XBlock.

In the process some deprecated functionality has been removed:
- Reading `data` and transforms being applied in the `init()` method.
- The `source` field.
- The `source_visible` attribute.

**Sandbox URL**: https://pr20384.sandbox.opencraft.hosting/ (the standard devstack accounts are enabled)

**Testing instructions**:

Test that editing and viewing of video xblocks work.

**OC Reviewers**
- [ ] @pomegranited

edx-webhook commented:
Thanks for the pull request, @symbolist! I've created [OSPR-3343](https://openedx.atlassian.net/browse/OSPR-3343) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

natabene commented:
@symbolist Thank you for your contribution. Please let me know once all checks are green.

symbolist commented:
jenkins run bokchoy

pomegranited commented on common/lib/xmodule/xmodule/tests/test_video.py:
```diff
@@ -394,74 +394,13 @@ def test_from_xml_when_handout_is_course_asset(self, course_id_string, expected_
-            'data': '',
-            'transcripts': {},
-        })
-
```
Why were the above tests removed?

symbolist replied:
I initially thought they were related to the `source` field but looking at it again, with a small change, we can retain the rest of the test. So I have added it back. Thanks for catching this!

pomegranited commented on common/lib/xmodule/xmodule/video_module/video_module.py:
```diff
@@ -585,18 +547,16 @@ def save_with_metadata(self, user):
-        else:
-            editable_fields.pop('source')
+        # Source is no longer supported.
+        editable_fields.pop('source')
```
Will let @ormsbee answer this one -- my opinion is yes.

ormsbee replied:
👍

symbolist replied:
I have removed the `source` field in https://github.com/edx/edx-platform/commit/2a740629fc66ae884362816a73f91c97628e1282 and squashed the commit.

symbolist commented on cms/djangoapps/contentstore/management/commands/tests/test_migrate_transcripts.py:
```diff
@@ -105,9 +106,9 @@ def setUp(self):
+                   start_time="1.0"
                    download_video="false"
-                   end_time="00:01:00">
+                   end_time="60.0">
```
Ah yes, this was a tricky and I tried a number of ways to get it to work.

The problem is that `deserialize_fields` gets [called](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/xml_module.py#L83) on this value when [setting](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/video_module/video_module.py#L958) it in the field data. However, `json.loads()` fails because it is not a valid json value so the value that gets put is a string and not a time object.

This causes this error when trying to save the XBlock:
```
lms/djangoapps/courseware/tests/test_video_handlers.py:142: in initialize_block
    super(BaseTestVideoXBlock, self).initialize_module(**kwargs)
lms/djangoapps/courseware/tests/helpers.py:81: in initialize_module
    self.item_descriptor = ItemFactory.create(**kwargs)
../venvs/edxapp/local/lib/python2.7/site-packages/factory/base.py:623: in create
    return cls._generate(True, attrs)
../venvs/edxapp/local/lib/python2.7/site-packages/factory/base.py:548: in _generate
    obj = cls._prepare(create, **attrs)
../venvs/edxapp/local/lib/python2.7/site-packages/factory/base.py:523: in _prepare
    return cls._create(model_class, *args, **kwargs)
common/lib/xmodule/xmodule/modulestore/tests/factories.py:389: in _create
    fields=kwargs,
lms/djangoapps/ccx/modulestore.py:212: in create_child
    user_id, parent_usage_key, block_type, block_id=block_id, fields=fields, **kwargs
common/lib/xmodule/xmodule/modulestore/mixed.py:93: in inner
    retval = func(field_decorator=strip_key_collection, *args, **kwargs)
common/lib/xmodule/xmodule/modulestore/mixed.py:113: in wrapper
    return func(*args, **kwargs)
common/lib/xmodule/xmodule/modulestore/mixed.py:775: in create_child
    return modulestore.create_child(user_id, parent_usage_key, block_type, block_id=block_id, fields=fields, **kwargs)
common/lib/xmodule/xmodule/modulestore/mongo/base.py:1452: in create_child
    xblock = self.create_item(user_id, parent_usage_key.course_key, block_type, block_id=block_id, **kwargs)
common/lib/xmodule/xmodule/modulestore/mongo/base.py:1426: in create_item
    xblock = self.create_xblock(runtime, course_key, block_type, block_id, **kwargs)
common/lib/xmodule/xmodule/modulestore/mongo/draft.py:337: in create_xblock
    runtime, course_key, block_type, block_id, fields, **kwargs
common/lib/xmodule/xmodule/modulestore/mongo/base.py:1402: in create_xblock
    xmodule.save()
../venvs/edxapp/local/lib/python2.7/site-packages/xblock/mixins.py:255: in save
    self.force_save_fields(fields_to_save)
../venvs/edxapp/local/lib/python2.7/site-packages/xblock/mixins.py:265: in force_save_fields
    fields_to_save_json[field.name] = field.to_json(self._field_data_cache[field.name])
```

As far as I understand this issue only impacts the tests because the `start` and `end` values when loaded from `data` are not considered changed and so not saved during `xmodule.save()`.

@ormsbee Do you have any thoughts on this?

ormsbee replied:
That is painful to read, but given the circumstances, I think what you're doing is fine.

symbolist commented:
Hi @natabenem, this is ready for a review by the edX team. My understanding is that @ormsbee will be reviewing this XModule conversion as well. Some very old backwards compatibility code has been removed as part of that but I'll let @ormsbee comment on whether it requires a conversation with product. Thanks!

natabene commented:
@ormsbee Could you give this a look?

edx-webhook commented:
Thanks for the pull request, @symbolist! I've created [OSPR-3518](https://openedx.atlassian.net/browse/OSPR-3518) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

ormsbee commented:
@pdpinch, @Colin-Fredericks: This PR converts VideoModule into an XBlock, and has advanced to the point where it's ready for testing. The sandbox linked in the PR description is live. Would you like to kick the tires a bit and try importing/exporting your interesting Video cases?

pdpinch commented:
Yes, I would like to try some importing and exporting. I'd like also to be reassured that these changes play well with video that's uploaded through the edx video pipeline -- veda/edx-val. I suspect that there are some complex dependencies, particularly around transcripts and so-called thumbnails.

Colin-Fredericks commented:
@ormsbee Thanks for the note! I'll definitely put aside some time for testing next Monday and Wednesday.

symbolist commented:
@pdpinch We do not have the edx pipeline setup on our sandboxes, so that testing can only happen once this PR is on edX stage. But there were not any changes related to transcripts or uploads so I do not anticipate issues related to those.

Colin-Fredericks commented:
@symbolist edX has a staging server?? When did this happen?

Colin-Fredericks commented:
Testing time!

As usual I'm pulling things from my [Tarballs for Testing](https://drive.google.com/drive/folders/1UjEnA4q9EpPHyIsw7ZzWAdDGVLDKaB97?usp=sharing) folder. Anyone who can't see it and wants access, just let me know.

**Grape Ape:** our kitchen sink course.
Uploads, unpacks, and verifies correctly.
Breaks at the "Updating Course" stage, with the ever-helpful "Error Importing Course." At this point this is the common behavior I've come to expect from Grape Ape, but it's still not a good thing.

**HX Boilerplate Course:** our tiny wireframe that becomes the base for new courses.
Uploads, unpacks, and verifies correctly.
Successfully updates!
Videos show up properly in course.
After re-exporting, the video folder is present, with the correct three videos in it.

**Justice:** an actual real course.
Uploads, unpacks, and verifies correctly.
Failed at the "Updating Course" stage with the usual error. This wasn't expected.

The "Updating" stage was slow for my imports, but I'm assuming that it's just running on a slower server instead of a "Throw all of Amazon at it" model.

ormsbee commented:
@Colin-Fredericks 
> edX has a staging server?? When did this happen?

We've had one since pretty much the beginning but I believe we discourage its use because the data is out of sync and often the configurations don't match our production environments well (especially as more configuration made its way into the database layer). I'll poke folks on our side on where we can test on the VAL side of things.

Colin-Fredericks commented:
> We've had one since pretty much the beginning but I believe we discourage its use because the data is out of sync and often the configurations don't match our production environments well (especially as more configuration made its way into the database layer). I'll poke folks on our side on where we can test on the VAL side of things.

Ah, ok. I do remember pushing courses to stage for testing, and then to prod, in the long-long-ago before Studio. I was just surprised because there have been multiple issues in the last year or so where HX has said "This could have been detected on a staging server" (like the day when every single assessment disappeared from every single course) and no one said "But we have one."

Anyway, I'll stop filling this PR with this conversation; my apologies.

symbolist commented:
@Colin-Fredericks Thanks for testing!

I have uploaded the two courses. For Justice, I just downloaded the archive in the drive and uploaded it and it worked for me. For Grape Ape, I had a copy from the CAPA conversion PR without the annotation xblocks (that were causing failure to import) plus I had to install the ubcpi xblock. Would you like to take a look at them?

Colin-Fredericks commented:
The UBCPI is probably also what prevented Justice from importing. After you got those in, they both seem to have a video folder with the correct number of videos.

**Climate and Health**
Uploads, unpacks, and verifies correctly.
Updates fine, exports fine.
Video folder present with correct number of videos.

I think it's working!

symbolist commented:
jenkins run bokchoy

symbolist commented:
@Colin-Fredericks Thanks for the testing!

@ormsbee There was only changed required. I have added it in https://github.com/edx/edx-platform/commit/2a740629fc66ae884362816a73f91c97628e1282 and squashed it up. I hope the commit message contains all the necessary details.

pdpinch commented:
@shellyu can you share some test courses with Usman?

ormsbee commented:
@nedbat, @feanil: After MIT has completed their testing and @pdpinch has given his 👍 , may I trouble one of you folks to hit the merge button on this (since I'm going to be out for a couple of weeks)? You'd also need to drop a note in the #support channel to let them know exactly when this is going out.

feanil commented:
@ormsbee will do.  @pdpinch let me @symbolist let me know when you believe this is ready to merge.  I'll wait for a note from both of you before landing this.

pdpinch commented:
@feanil @symbolist some folks at MIT (@kboots-mit and @sfrucht) tested with 22.011x, 6.041x and 18.01x and they all imported and exported as expected.  

22.011x uses the VEDA workflow and its cloudfront-hosted videos won't play in the sandbox. But we expected that. 

6.041x and 18.01x are both written in XML and use YouTube videos. They worked without issue.

pdpinch commented:
To be explicit @feanil I believe this is ready to merge.

feanil commented:
@symbolist is this ready to merge from your perspective as well?

symbolist commented:
@pdpinch Thanks for the testing round!

@feanil Yup, this is ready for merging. Thanks!

edx-webhook commented:
@symbolist 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/20384&entry.1484655318=https://github.com/symbolist&entry.752974735=2019-05-02+17:23&entry.1917517419=2019-06-27+20:23&entry.2133058324=Yes) so we can improve your experience in the future.
